src/ray/raylet/scheduling/policy/node_label_scheduling_policy.cc

```diff
@@ -0,0 +1,189 @@
+// Copyright 2023 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/raylet/scheduling/policy/node_label_scheduling_policy.h"
+
+#include "ray/raylet/scheduling/policy/scheduling_context.h"
+
+namespace ray {
+namespace raylet_scheduling_policy {
+
+scheduling::NodeID NodeLabelSchedulingPolicy::Schedule(
+    const ResourceRequest &resource_request, SchedulingOptions options) {
+  RAY_CHECK(options.scheduling_type == SchedulingType::NODE_LABEL);
+  auto context =
+      dynamic_cast<const NodeLabelSchedulingContext *>(options.scheduling_context.get());
+  const auto &scheduling_strategy = context->GetSchedulingStrategy();
+  RAY_CHECK(scheduling_strategy.has_node_label_scheduling_strategy());
+  const auto &node_label_scheduling_strategy =
+      scheduling_strategy.node_label_scheduling_strategy();
+  auto candidate_nodes = SelectAvailableNode(resource_request);
+  if (candidate_nodes.empty()) {
+    return scheduling::NodeID::Nil();
```

Comment:
Returning Nil means that there is no feasible node in the cluster for this resource request, which is not what we want.

I think the overall logic should be:

1. If we can find available nodes that match the hard and soft expressions, return a random one.
2. If we can find available nodes that match the hard expression but not soft one, return a random one.
3. If we can find feasible nodes that match the hard and soft expressions, return a random one.
4. If we can find feasible nodes that match the hard expression but not soft one, return a random one.
5. Return Nil.

Reply:
Let me take some time to think about it.

Reply:
Fixed.